planet/data_filter.py

```diff
@@ -0,0 +1,118 @@
+# the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+"""Functionality for preparing a data search filter"""
+from datetime import datetime
+import logging
+from typing import List, Union
+
+from planet import exceptions
+
+LOGGER = logging.getLogger(__name__)
+
+
+def and_filter(nested_filters: List[dict]) -> dict:
+    """Create an AndFilter
+
+    The AndFilter can be used to limit results to items with properties or
+    permissions which match all nested filters.
+
+    It is most commonly used as a top-level filter to ensure criteria across
+    all field and permission filters are met.
+
+    Parameters:
+        nested_filters: Filters to AND together
+    """
+    return {'type': 'AndFilter', 'config': nested_filters}
+
+
+def or_filter(nested_filters: List[dict]) -> dict:
+    """Create an OrFilter
+
+    The OrFilter can be used to match items with properties or permissions
+    which match at least one of the nested filters.
+
+    Parameters:
+        nested_filters: Filters to OR together
+    """
+    return {'type': 'OrFilter', 'config': nested_filters}
+
+
+def not_filter(nested_filter: dict):
+    """Create a NotFilter
+
+    The NotFilter can be used to match items with properties or permissions
+    which do not match the nested filter.
+
+    This filter only supports a single nested filter. Multiple NotFilters can
+    be nested within an AndFilter to filter across multiple fields or
+    permission values.
+
+    Parameters:
+        nested_filter: Filter to NOT
+    """
+    return {'type': 'NotFilter', 'config': nested_filter}
+
+
+def _field_filter(ftype: str, field_name: str, config: Union[dict,
+                                                             list]) -> dict:
+    return {'type': ftype, 'field_name': field_name, 'config': config}
+
+
+def date_range_filter(field_name: str,
+                      gt: datetime = None,
+                      lt: datetime = None,
+                      gte: datetime = None,
+                      lte: datetime = None) -> dict:
+    """Create a DateRangeFilter
+
+    The DateRangeFilter can be used to search on any property with a timestamp
+    such as acquired or published.
+
+    One or more of the conditional parameters `gt`, `lt`, `gte`, `lte` must be
+    specified.
+
+    Parameters:
+        field_name: Name of field to filter on
+        gt: Filter to field timestamp later than this value
+        lt: Filter to field timestamp earlier than this value
+        gte: Filter to field timestamp at or later than this value
+        lte: Filter to field timestamp at or earlier than this value
```

Comment:
@jreiberkyle is the result the logical conjunction of all of these filters, or the disjunction? I wonder if the doc should say something about that.

Reply:
For example, what will be the result of `date_range_filter(lt=x, gt=x)`?

Reply:
yeah it's a good idea to be explicit. fyi in response to your question, that query would get no results.